oss-internship-2020/guetzli/guetzli_transaction.cc

```diff
@@ -0,0 +1,144 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "guetzli_transaction.h"
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <iostream>
+#include <memory>
+
+namespace guetzli {
+namespace sandbox {
+
+absl::Status GuetzliTransaction::Main() {
+  sapi::v::Fd in_fd(open(params_.in_file, O_RDONLY));
+
+  if (in_fd.GetValue() < 0) {
+    return absl::FailedPreconditionError(
+      "Error opening input file"
+    );
+  }
+
+  SAPI_ASSIGN_OR_RETURN(image_type_, GetImageTypeFromFd(in_fd.GetValue()));
+  SAPI_RETURN_IF_ERROR(sandbox()->TransferToSandboxee(&in_fd));
+
+  if (in_fd.GetRemoteFd() < 0) {
+    return absl::FailedPreconditionError(
+        "Error receiving remote FD: remote input fd is set to -1");
+  }
+
+  GuetzliApi api(sandbox());
+  sapi::v::LenVal output(0);
+
+  sapi::v::Struct<ProcessingParams> processing_params;
+  *processing_params.mutable_data() = {in_fd.GetRemoteFd(), 
+                                      params_.verbose, 
+                                      params_.quality, 
+                                      params_.memlimit_mb
+  };
+
+  auto result = image_type_ == ImageType::kJpeg ? 
+    api.ProcessJpeg(processing_params.PtrBefore(), output.PtrBoth()) : 
```

Comment:
`PtrBefore` can be used for output and then `PtrNone` for `WriteDataToFd`.
That way the data will stay inside the sandboxee and will just get written to the output file.
Otherwise it is copied twice - once after `ProcessJpeg` call and then once again before `WriteDataToFd`